### ISSUE TITLE
no-fallthrough doesn't seem to work

### ISSUE THREAD

**Issue by Grandrath** [eslint/eslint#5799]
Maybe I am missing something, but the following code snippet lints successfully when I enter it into http://eslint.org/demo/. I expected it to fail with `no-fallthrough`.

```javascript
switch ("foo") {
case "foo":
case "bar":
    break;
default:
    throw new Error();
}
```

**Comment by scriptdaemon** [comment]
This rule allows for empty case clauses, which explains why that snippet of code that does not generate a warning. A similar snippet is included among the examples of passing code for this rule: http://eslint.org/docs/rules/no-fallthrough.

**Comment by Grandrath** [comment]
Thank you for the clarification. Do you think it might be useful for others to add an option (e.g. `AllowEmptyCase`) to change this behavior? In a current project of mine a fellow developer commented out the content of a `case` clause. This didn't cause a bug in this particular instance but it would have been good if ESLint caught this right away.

**Comment by platinumazure** [comment]
I think it would be good to return a lint error in this case:

```js
switch (something) {
    case 1:
        // commented out code, unintentional fallthrough
    case 2:
        somethingElse();
        break;
}
```

Basically, I think an argument could be made that line 3 should be warned (i.e., any comment between two case labels, which otherwise represents an empty case block, should not be considered fully empty for the purpose of this rule-- unless the comment has "falls through" or whatever other language in it).

**Comment by ilyavolodin** [comment]
Code example that you provided would be an enhancement, and we would need to put it behind an option.

**Comment by platinumazure** [comment]
@Grandrath To be clear, are you requesting an option to enable lint errors for all empty cases, or just ones that are empty syntactically but which have comments that don't match the "falls through" expression?

Put another way, should this warn on the `case 1` line?

```js
switch (something) {
    case 1:
    case 2:
        yay();
}
```

**Comment by nzakas** [comment]
We originally had this as an exception because it's a pretty common way to say you want fall through (even JSLint allows it):

```js
switch (something) {
    case 1:
    case 2:
        yay();
}
```

My opinion is that this should still not warn. However, this definitely should warn:

```js
switch (something) {
    case 1:
        // yay();
    case 2:
        yay();
}
```

So, if I've read this issue correct, it's really the latter case that is problematic and I'd say it's a bug that it's not being flagged right now. @Grandrath is that correct?

**Comment by Grandrath** [comment]
@nzakas Yes, that is correct.

**Comment by spadgos** [comment]
This would be quite strange to me. What about code like this which uses comments to document the cases?

```js
switch (state) {
  case 1: // connecting
  case 2: // open
    doSomething();
}
```

**Comment by michaelficarra** [comment]
@spadgos That would be better written without the magic numbers to begin with:

```js
switch (state) {
  case CONNECTING:
  case OPEN:
    doSomething();
}
```

**Comment by nzakas** [comment]
@Grandrath thanks. Do you want to submit a PR?

**Comment by alberto** [comment]
Mmm, I tried to submit a PR and it failed in a couple of places in our own codebase because of the concerns expressed by @spadgos 
From `lib/rules/valid-jsdoc.js`:
```
case "UnionType":  // {String|number|Test}
case "ArrayType":  // {[String, number, Test]}
    elements = type.elements;
    break;
```

From `lib/rules/no-mixed-spaces-and-tabs.js`:
```
case true: // Support old syntax, maybe add deprecation warning here
case "smart-tabs":
```

**Comment by pedrottimark** [comment]
Yeah, even though there is a better way https://github.com/eslint/eslint/issues/5799#issuecomment-207466093 haven’t most of us done it? It feels like a breaking change.

The following code passes JSLint with default options:

```
'use strict';
function f(n) {
    switch (n) {
    case 0: // zero means nothing
    case 1: // one means one thing
        return 'thing or nothing';
    default:
        return 'things';
    }
}```

**Comment by nzakas** [comment]
The only time it should fail is when there is an extra line between the cases. This should not fall:

```js
case "UnionType":  // {String|number|Test}
case "ArrayType":  // {[String, number, Test]}
    elements = type.elements;
    break;
```

This should fail:

```js
case "UnionType":
    // {String|number|Test}
case "ArrayType":  // {[String, number, Test]}
    elements = type.elements;
    break;
```

**Comment by ilyavolodin** [comment]
Hmm I'm curious should this example fail?
```js
case "UnitonType":

case "ArrayType":
  elements = type.elements;
  break;
```

**Comment by alberto** [comment]
I think he meant the comment starts in a different line

**Comment by alberto** [comment]
I updated the PR to reflect that, let me know if you don't think this is the right solution.

**Comment by nzakas** [comment]
@ilyavolodin yes it should. 

@alberto I meant whenever cases are more than one line apart, whether because of empty lines or comments, it should warn. So @ilyavolodin's example should warn.

**Comment by alberto** [comment]
Should this warn too?
```
case "UnitonType":  /* Some long comment
                       spanning multiple lines */
case "ArrayType":
  elements = type.elements;
  break;
```

**Comment by nzakas** [comment]
Yup. If the two cases aren't directly adjacent, then it should warn.
